content/posts/2025-01-03-packets-xlation-enrichment.md

```diff
@@ -0,0 +1,197 @@
+---
+layout: :theme/post
+title: "Enhancing NetObserv for Kubernetes Service Flows using eBPF"
+description: NetObserv Packet Translation provide enriched service endpoint POD information
+tags: eBPF,Services,Monitoring,Troubleshooting
+authors: [msherif1234]
+---
+
+# Enhancing NetObserv for Kubernetes Service Flows using eBPF: Adding Translated Endpoint Information
+
+## Introduction
+
+In a cloud-native environment, Kubernetes has become the de facto standard for managing containerized workloads.
+While its service discovery and load-balancing features are robust, gaining visibility into the actual endpoints
+serving traffic often requires complex instrumentation or external tooling.
+This is where eBPF (extended Berkeley Packet Filter) shines, offering a way to deeply observe and enrich Kubernetes
+service flows without intrusive changes to your applications.
+
+In this blog, we'll explore how eBPF can be used to add translated endpoint (xlated endpoint) information to
+Kubernetes service flows, providing insights into backend behavior and improving observability.
+
+## The Challenge
+
+When traffic flows through a Kubernetes Service, it often gets abstracted away by kube-proxy or other load balancers.
+For example, a request to `my-service` on `ClusterIP` is transparently routed to one of the available pods.
+However, network observability will capture both the service traffic and the virtual service endpoint as two separate flows.
+This lack of granularity makes troubleshooting and optimization more challenging.
+
+To solve this, we can:
+
+Capture the network flows at the kernel level.
+
+Enrich them with translated endpoint information, showing not only the service but also the specific backend pod.
```

Comment:
Could you please add a quick drawing explaining this ? I guess it will help :smile:

Reply:
the ui output with the svc traffic example not enough ?

Reply:
added new diagram to show how it works let me know what u think

Reply:
thanks !